rfcs/text/0003_application_service_mounting.md

```diff
@@ -0,0 +1,259 @@
+- Start Date: 2019-05-10
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+A front-end service to manage registration and root-level routing for
+first-class applications.
+
+# Basic example
+
+```tsx
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import { MyApp } from './components';
+
+class MyPlugin {
+  setup({ application }) {
+    application.registerApp({
+      id: 'my-app',
+      title: 'My Application',
+      rootRoute: '/myapp',
+      mount(targetDomElement, pluginStartContext, pluginStart) {
+        ReactDOM.render(
+          <MyAppRoot core={pluginStartContext} deps={pluginStart} />,
+          targetDomElement
+        );
+
+        return () => {
+          ReactDOM.unmountComponentAtNode(targetDomElement);
+        };
+      }
+    });
+  }
+}
+```
+
+# Motivation
+
+By having centralized management of applications we can have a true single page
+application. It also gives us a single place to enforce authorization and/or
+licensing constraints on application access.
+
+By making the mounting interface of the ApplicationService generic, we can
+support many different rendering technologies simultaneously to avoid framework
+lock-in.
+
+# Detailed design
+
+This is the bulk of the RFC. Explain the design in enough detail for somebody
+familiar with Kibana to understand, and for somebody familiar with the
+implementation to implement. This should get into specifics and corner-cases,
+and include examples of how the feature is used. Any new terminology should be
+defined here.
+
+## Interface
+
+```ts
+export type Unmount = () => void;
```

Comment:
An application might wish to send unsaved state to the server before completing the unmount, so I think this should return a `Promise`

Reply:
I think this is a great point. A situation where you might want to track UI metric telemetry related to unmounting comes to mind. But why would we want to block unmounting on the resolution of a request? I can't think of a scenario where we'd want to prevent or delay an unmount since it would create a strange UX.

Reply:
I agree with CJ that unmount should block the transition, and I also think we should call unmount on page refresh and such (window’s unload event) in which case we can’t wait for async handlers.